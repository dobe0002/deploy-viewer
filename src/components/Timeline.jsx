/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react';
import PropTypes from 'prop-types';
import map from 'lodash/map';
import styled from 'styled-components';
import { Container, Row, Col, Table } from 'reactstrap';

import TimelineRow from './TimelineRow';
import TimelineHeaderRow from './TimelineHeaderRow';
import FilterDateAppGroup from './FilterDateAppGroup';
import { getStartEndDate, getDates, convertDateToString } from '../js/getDates';
import 'react-datepicker/dist/react-datepicker.min.css';

export default class Timeline extends React.Component {
  constructor(props) {
    super(props);
    const { deployData, currentAppGroup, startDate, endDate } = this.props;
    this.setAppRows = this.setAppRows.bind(this);
    this.handleStartChange = this.handleStartChange.bind(this);
    this.handleEndChange = this.handleEndChange.bind(this);
    this.handleAppGroupChange = this.handleAppGroupChange.bind(this);
    this.setHistory = this.setHistory.bind(this);

    const dateRange = getStartEndDate(deployData);

    const sDate = startDate || dateRange.min;
    const eDate = endDate || dateRange.max;
    this.state = {
      startDate: sDate,
      endDate: eDate,
      dates: getDates(sDate, eDate),
      currentAppGroup
    };
  }

  setAppRows() {
    const { deployData } = this.props;
    const { dates, currentAppGroup } = this.state;
    return map(deployData, app => {
      if (currentAppGroup === '' || currentAppGroup === app.appGroup) {
        return (
          <TimelineRow
            key={app.appCode}
            dateArray={dates}
            appStatusArray={app.status}
            appTitle={app.title}
          />
        );
      }
      return '';
    });
  }

  setHistory() {
    const { history } = this.props;
    const { startDate, endDate, currentAppGroup } = this.state;
    const sDate = convertDateToString(startDate);
    const eDate = convertDateToString(endDate);

    history.push(
      `/timeline/${currentAppGroup}/startDate/${sDate}/endDate/${eDate}`
    );
  }

  handleAppGroupChange(newAppGroup) {
    // const { history } = this.props;
    this.setState({ currentAppGroup: newAppGroup }, () => {
      this.setHistory();
    });
  }

  handleStartChange(date) {
    const { endDate } = this.state;
    const newDates = getDates(date, endDate);
    this.setState(
      {
        startDate: date,
        dates: newDates
      },
      () => {
        this.setHistory();
      }
    );
  }

  handleEndChange(date) {
    const { startDate } = this.state;
    const newDates = getDates(startDate, date);
    this.setState(
      {
        endDate: date,
        dates: newDates
      },
      () => {
        this.setHistory();
      }
    );
  }

  render() {
    const TableDiv = styled.div`
      td,
      th {
        padding: 0.25rem;
      }
      tbody {
        td {
          text-align: center;
        }
      }
      thead {
        td {
          font-weight: bold;
        }
        th {
          font-weight: normal;
        }
      }
    `;

    const { startDate, endDate, dates, currentAppGroup } = this.state;
    const { appGroups } = this.props;
    return (
      <Container fluid>
        <FilterDateAppGroup
          dataTitle='Deploy Timeline'
          startDate={startDate}
          endDate={endDate}
          currentAppGroup={currentAppGroup}
          appGroups={appGroups}
          onStartDateChange={this.handleStartChange}
          onEndDateChange={this.handleEndChange}
          onAppGroupChange={this.handleAppGroupChange}
        />
        <Row>
          <Col>
            <TableDiv>
              <Table bordered striped hover>
                <TimelineHeaderRow dateArray={dates} />
                <tbody>{this.setAppRows()}</tbody>
              </Table>
            </TableDiv>
          </Col>
        </Row>
      </Container>
    );
  }
}

Timeline.propTypes = {
  deployData: PropTypes.arrayOf.isRequired,
  appGroups: PropTypes.objectOf,
  currentAppGroup: PropTypes.string,
  history: PropTypes.objectOf,
  startDate: PropTypes.instanceOf(Date),
  endDate: PropTypes.instanceOf(Date)
};

Timeline.defaultProps = {
  appGroups: {},
  currentAppGroup: '',
  history: { push: () => {} },
  startDate: undefined,
  endDate: undefined
};
