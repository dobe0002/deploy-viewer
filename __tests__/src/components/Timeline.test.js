/* eslint-disable react/jsx-filename-extension */
import React from 'react';

import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Timeline from '../../../src/components/Timeline';
import deployData from '../../../__fixtures__/getDataFormat/sheetData';

configure({ adapter: new Adapter() });

const { axe, toHaveNoViolations } = require('jest-axe');

expect.extend(toHaveNoViolations);

describe('Timeline component tests', () => {
  const sampleProps = {
    deployData
  };

  test('Accessibility check', async () => {
    const div = document.createElement('div');
    div.id = 'popover2';
    document.body.appendChild(div);
    const wrapper = shallow(<Timeline deployData={sampleProps.deployData} />);
    const results = await axe(wrapper.html());
    expect(results).toHaveNoViolations();
  });
  test('Snapshot test', async () => {
    const div = document.createElement('div');
    div.id = 'popover2';
    document.body.appendChild(div);
    const wrapper = shallow(<Timeline deployData={sampleProps.deployData} />);
    expect(wrapper.find('TimelineRow')).toHaveLength(4);
    expect(wrapper).toMatchSnapshot();
  });
  test('Change app group', () => {
    const div = document.createElement('div');
    div.id = 'popover2';
    document.body.appendChild(div);
    const wrapper = shallow(<Timeline deployData={sampleProps.deployData} />);

    expect(wrapper.state('currentAppGroup')).toEqual('');
    expect(wrapper.find('TimelineRow')).toHaveLength(4);

    wrapper.instance().handleAppGroupChange('HR');
    expect(wrapper.state('currentAppGroup')).toEqual('HR');
    expect(wrapper.find('TimelineRow')).toHaveLength(2);
  });
  test('Change start and end dates', () => {
    const div = document.createElement('div');
    div.id = 'popover2';
    document.body.appendChild(div);
    const wrapper = shallow(<Timeline deployData={sampleProps.deployData} />);

    const startDate = new Date('2019-11-01T06:00:00.000Z');
    const endDate = new Date('2019-12-15T06:00:00.000Z');
    const newStartDate = new Date(2019, 10, 15);
    const newEndDate = new Date(2019, 11, 31);

    expect(wrapper.state('startDate')).toEqual(startDate);
    expect(wrapper.state('endDate')).toEqual(endDate);

    wrapper.instance().handleStartChange(newStartDate);
    expect(wrapper.state('startDate')).toEqual(newStartDate);
    expect(wrapper.state('endDate')).toEqual(endDate);

    wrapper.instance().handleEndChange(newEndDate);
    expect(wrapper.state('startDate')).toEqual(newStartDate);
    expect(wrapper.state('endDate')).toEqual(newEndDate);
  });
});
