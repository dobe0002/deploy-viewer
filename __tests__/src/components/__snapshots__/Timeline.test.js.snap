// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline component tests Snapshot test 1`] = `
<Container
  fluid={true}
  tag="div"
>
  <FilterDateAppGroup
    appGroups={Object {}}
    currentAppGroup=""
    dataTitle="Deploy Timeline"
    endDate={2019-12-15T06:00:00.000Z}
    onAppGroupChange={[Function]}
    onEndDateChange={[Function]}
    onStartDateChange={[Function]}
    startDate={2019-11-01T05:00:00.000Z}
  />
  <Row
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <styled.div>
        <Table
          bordered={true}
          hover={true}
          responsiveTag="div"
          striped={true}
          tag="table"
        >
          <TimelineHeaderRow
            dateArray={
              Array [
                2019-11-01T05:00:00.000Z,
                2019-11-02T05:00:00.000Z,
                2019-11-03T05:00:00.000Z,
                2019-11-04T06:00:00.000Z,
                2019-11-05T06:00:00.000Z,
                2019-11-06T06:00:00.000Z,
                2019-11-07T06:00:00.000Z,
                2019-11-08T06:00:00.000Z,
                2019-11-09T06:00:00.000Z,
                2019-11-10T06:00:00.000Z,
                2019-11-11T06:00:00.000Z,
                2019-11-12T06:00:00.000Z,
                2019-11-13T06:00:00.000Z,
                2019-11-14T06:00:00.000Z,
                2019-11-15T06:00:00.000Z,
                2019-11-16T06:00:00.000Z,
                2019-11-17T06:00:00.000Z,
                2019-11-18T06:00:00.000Z,
                2019-11-19T06:00:00.000Z,
                2019-11-20T06:00:00.000Z,
                2019-11-21T06:00:00.000Z,
                2019-11-22T06:00:00.000Z,
                2019-11-23T06:00:00.000Z,
                2019-11-24T06:00:00.000Z,
                2019-11-25T06:00:00.000Z,
                2019-11-26T06:00:00.000Z,
                2019-11-27T06:00:00.000Z,
                2019-11-28T06:00:00.000Z,
                2019-11-29T06:00:00.000Z,
                2019-11-30T06:00:00.000Z,
                2019-12-01T06:00:00.000Z,
                2019-12-02T06:00:00.000Z,
                2019-12-03T06:00:00.000Z,
                2019-12-04T06:00:00.000Z,
                2019-12-05T06:00:00.000Z,
                2019-12-06T06:00:00.000Z,
                2019-12-07T06:00:00.000Z,
                2019-12-08T06:00:00.000Z,
                2019-12-09T06:00:00.000Z,
                2019-12-10T06:00:00.000Z,
                2019-12-11T06:00:00.000Z,
                2019-12-12T06:00:00.000Z,
                2019-12-13T06:00:00.000Z,
                2019-12-14T06:00:00.000Z,
                2019-12-15T06:00:00.000Z,
              ]
            }
          />
          <tbody>
            <TimelineRow
              appStatusArray={
                Array [
                  Object {
                    "date": 2019-12-01T06:00:00.000Z,
                    "dateSort": 20191201,
                    "dateString": "12/01/2019",
                    "note": "Users could not log in",
                    "row": 10,
                    "status": "C",
                    "statusDescription": "Deployment with Customer Impact",
                  },
                  Object {
                    "date": 2019-12-01T06:00:00.000Z,
                    "dateSort": 20191201,
                    "dateString": "12/01/2019",
                    "note": "Fixed issues from the morning deployment",
                    "row": 11,
                    "status": "G",
                    "statusDescription": "Good deployment",
                  },
                  Object {
                    "date": 2019-12-15T06:00:00.000Z,
                    "dateSort": 20191215,
                    "dateString": "12/15/2019",
                    "note": ":-)",
                    "row": 9,
                    "status": "G",
                    "statusDescription": "Good deployment",
                  },
                ]
              }
              appTitle="Cap"
              dateArray={
                Array [
                  2019-11-01T05:00:00.000Z,
                  2019-11-02T05:00:00.000Z,
                  2019-11-03T05:00:00.000Z,
                  2019-11-04T06:00:00.000Z,
                  2019-11-05T06:00:00.000Z,
                  2019-11-06T06:00:00.000Z,
                  2019-11-07T06:00:00.000Z,
                  2019-11-08T06:00:00.000Z,
                  2019-11-09T06:00:00.000Z,
                  2019-11-10T06:00:00.000Z,
                  2019-11-11T06:00:00.000Z,
                  2019-11-12T06:00:00.000Z,
                  2019-11-13T06:00:00.000Z,
                  2019-11-14T06:00:00.000Z,
                  2019-11-15T06:00:00.000Z,
                  2019-11-16T06:00:00.000Z,
                  2019-11-17T06:00:00.000Z,
                  2019-11-18T06:00:00.000Z,
                  2019-11-19T06:00:00.000Z,
                  2019-11-20T06:00:00.000Z,
                  2019-11-21T06:00:00.000Z,
                  2019-11-22T06:00:00.000Z,
                  2019-11-23T06:00:00.000Z,
                  2019-11-24T06:00:00.000Z,
                  2019-11-25T06:00:00.000Z,
                  2019-11-26T06:00:00.000Z,
                  2019-11-27T06:00:00.000Z,
                  2019-11-28T06:00:00.000Z,
                  2019-11-29T06:00:00.000Z,
                  2019-11-30T06:00:00.000Z,
                  2019-12-01T06:00:00.000Z,
                  2019-12-02T06:00:00.000Z,
                  2019-12-03T06:00:00.000Z,
                  2019-12-04T06:00:00.000Z,
                  2019-12-05T06:00:00.000Z,
                  2019-12-06T06:00:00.000Z,
                  2019-12-07T06:00:00.000Z,
                  2019-12-08T06:00:00.000Z,
                  2019-12-09T06:00:00.000Z,
                  2019-12-10T06:00:00.000Z,
                  2019-12-11T06:00:00.000Z,
                  2019-12-12T06:00:00.000Z,
                  2019-12-13T06:00:00.000Z,
                  2019-12-14T06:00:00.000Z,
                  2019-12-15T06:00:00.000Z,
                ]
              }
              key="CAP"
            />
            <TimelineRow
              appStatusArray={
                Array [
                  Object {
                    "date": 2019-11-15T06:00:00.000Z,
                    "dateSort": 20191115,
                    "dateString": "11/15/2019",
                    "note": "Let's try some \\"BOLD\\" text",
                    "row": 8,
                    "status": "C",
                    "statusDescription": "Deployment with Customer Impact",
                  },
                ]
              }
              appTitle="Ethos"
              dateArray={
                Array [
                  2019-11-01T05:00:00.000Z,
                  2019-11-02T05:00:00.000Z,
                  2019-11-03T05:00:00.000Z,
                  2019-11-04T06:00:00.000Z,
                  2019-11-05T06:00:00.000Z,
                  2019-11-06T06:00:00.000Z,
                  2019-11-07T06:00:00.000Z,
                  2019-11-08T06:00:00.000Z,
                  2019-11-09T06:00:00.000Z,
                  2019-11-10T06:00:00.000Z,
                  2019-11-11T06:00:00.000Z,
                  2019-11-12T06:00:00.000Z,
                  2019-11-13T06:00:00.000Z,
                  2019-11-14T06:00:00.000Z,
                  2019-11-15T06:00:00.000Z,
                  2019-11-16T06:00:00.000Z,
                  2019-11-17T06:00:00.000Z,
                  2019-11-18T06:00:00.000Z,
                  2019-11-19T06:00:00.000Z,
                  2019-11-20T06:00:00.000Z,
                  2019-11-21T06:00:00.000Z,
                  2019-11-22T06:00:00.000Z,
                  2019-11-23T06:00:00.000Z,
                  2019-11-24T06:00:00.000Z,
                  2019-11-25T06:00:00.000Z,
                  2019-11-26T06:00:00.000Z,
                  2019-11-27T06:00:00.000Z,
                  2019-11-28T06:00:00.000Z,
                  2019-11-29T06:00:00.000Z,
                  2019-11-30T06:00:00.000Z,
                  2019-12-01T06:00:00.000Z,
                  2019-12-02T06:00:00.000Z,
                  2019-12-03T06:00:00.000Z,
                  2019-12-04T06:00:00.000Z,
                  2019-12-05T06:00:00.000Z,
                  2019-12-06T06:00:00.000Z,
                  2019-12-07T06:00:00.000Z,
                  2019-12-08T06:00:00.000Z,
                  2019-12-09T06:00:00.000Z,
                  2019-12-10T06:00:00.000Z,
                  2019-12-11T06:00:00.000Z,
                  2019-12-12T06:00:00.000Z,
                  2019-12-13T06:00:00.000Z,
                  2019-12-14T06:00:00.000Z,
                  2019-12-15T06:00:00.000Z,
                ]
              }
              key="ETHOS"
            />
            <TimelineRow
              appStatusArray={
                Array [
                  Object {
                    "date": 2019-11-02T05:00:00.000Z,
                    "dateSort": 20191102,
                    "dateString": "11/02/2019",
                    "note": "Hooray everything went to plan",
                    "row": 3,
                    "status": "G",
                    "statusDescription": "Good deployment",
                  },
                  Object {
                    "date": 2019-11-10T06:00:00.000Z,
                    "dateSort": 20191110,
                    "dateString": "11/10/2019",
                    "note": "People are trashing the database",
                    "row": 5,
                    "status": "O",
                    "statusDescription": "Outage",
                  },
                  Object {
                    "date": 2019-12-01T06:00:00.000Z,
                    "dateSort": 20191201,
                    "dateString": "12/01/2019",
                    "note": "This is a note with

Line breaks",
                    "row": 6,
                    "status": "G",
                    "statusDescription": "Good deployment",
                  },
                  Object {
                    "date": 2019-12-02T06:00:00.000Z,
                    "dateSort": 20191202,
                    "dateString": "12/02/2019",
                    "note": "",
                    "row": 7,
                    "status": "C",
                    "statusDescription": "Deployment with Customer Impact",
                  },
                ]
              }
              appTitle="Picker"
              dateArray={
                Array [
                  2019-11-01T05:00:00.000Z,
                  2019-11-02T05:00:00.000Z,
                  2019-11-03T05:00:00.000Z,
                  2019-11-04T06:00:00.000Z,
                  2019-11-05T06:00:00.000Z,
                  2019-11-06T06:00:00.000Z,
                  2019-11-07T06:00:00.000Z,
                  2019-11-08T06:00:00.000Z,
                  2019-11-09T06:00:00.000Z,
                  2019-11-10T06:00:00.000Z,
                  2019-11-11T06:00:00.000Z,
                  2019-11-12T06:00:00.000Z,
                  2019-11-13T06:00:00.000Z,
                  2019-11-14T06:00:00.000Z,
                  2019-11-15T06:00:00.000Z,
                  2019-11-16T06:00:00.000Z,
                  2019-11-17T06:00:00.000Z,
                  2019-11-18T06:00:00.000Z,
                  2019-11-19T06:00:00.000Z,
                  2019-11-20T06:00:00.000Z,
                  2019-11-21T06:00:00.000Z,
                  2019-11-22T06:00:00.000Z,
                  2019-11-23T06:00:00.000Z,
                  2019-11-24T06:00:00.000Z,
                  2019-11-25T06:00:00.000Z,
                  2019-11-26T06:00:00.000Z,
                  2019-11-27T06:00:00.000Z,
                  2019-11-28T06:00:00.000Z,
                  2019-11-29T06:00:00.000Z,
                  2019-11-30T06:00:00.000Z,
                  2019-12-01T06:00:00.000Z,
                  2019-12-02T06:00:00.000Z,
                  2019-12-03T06:00:00.000Z,
                  2019-12-04T06:00:00.000Z,
                  2019-12-05T06:00:00.000Z,
                  2019-12-06T06:00:00.000Z,
                  2019-12-07T06:00:00.000Z,
                  2019-12-08T06:00:00.000Z,
                  2019-12-09T06:00:00.000Z,
                  2019-12-10T06:00:00.000Z,
                  2019-12-11T06:00:00.000Z,
                  2019-12-12T06:00:00.000Z,
                  2019-12-13T06:00:00.000Z,
                  2019-12-14T06:00:00.000Z,
                  2019-12-15T06:00:00.000Z,
                ]
              }
              key="PIC"
            />
            <TimelineRow
              appStatusArray={
                Array [
                  Object {
                    "date": 2019-11-01T05:00:00.000Z,
                    "dateSort": 20191101,
                    "dateString": "11/01/2019",
                    "note": "Ooopsie!",
                    "row": 2,
                    "status": "B",
                    "statusDescription": "Bad Deployment",
                  },
                  Object {
                    "date": 2019-11-10T06:00:00.000Z,
                    "dateSort": 20191110,
                    "dateString": "11/10/2019",
                    "note": "Some stuff happened but it was ALL GOOD",
                    "row": 4,
                    "status": "G",
                    "statusDescription": "Good deployment",
                  },
                ]
              }
              appTitle="Training Hub"
              dateArray={
                Array [
                  2019-11-01T05:00:00.000Z,
                  2019-11-02T05:00:00.000Z,
                  2019-11-03T05:00:00.000Z,
                  2019-11-04T06:00:00.000Z,
                  2019-11-05T06:00:00.000Z,
                  2019-11-06T06:00:00.000Z,
                  2019-11-07T06:00:00.000Z,
                  2019-11-08T06:00:00.000Z,
                  2019-11-09T06:00:00.000Z,
                  2019-11-10T06:00:00.000Z,
                  2019-11-11T06:00:00.000Z,
                  2019-11-12T06:00:00.000Z,
                  2019-11-13T06:00:00.000Z,
                  2019-11-14T06:00:00.000Z,
                  2019-11-15T06:00:00.000Z,
                  2019-11-16T06:00:00.000Z,
                  2019-11-17T06:00:00.000Z,
                  2019-11-18T06:00:00.000Z,
                  2019-11-19T06:00:00.000Z,
                  2019-11-20T06:00:00.000Z,
                  2019-11-21T06:00:00.000Z,
                  2019-11-22T06:00:00.000Z,
                  2019-11-23T06:00:00.000Z,
                  2019-11-24T06:00:00.000Z,
                  2019-11-25T06:00:00.000Z,
                  2019-11-26T06:00:00.000Z,
                  2019-11-27T06:00:00.000Z,
                  2019-11-28T06:00:00.000Z,
                  2019-11-29T06:00:00.000Z,
                  2019-11-30T06:00:00.000Z,
                  2019-12-01T06:00:00.000Z,
                  2019-12-02T06:00:00.000Z,
                  2019-12-03T06:00:00.000Z,
                  2019-12-04T06:00:00.000Z,
                  2019-12-05T06:00:00.000Z,
                  2019-12-06T06:00:00.000Z,
                  2019-12-07T06:00:00.000Z,
                  2019-12-08T06:00:00.000Z,
                  2019-12-09T06:00:00.000Z,
                  2019-12-10T06:00:00.000Z,
                  2019-12-11T06:00:00.000Z,
                  2019-12-12T06:00:00.000Z,
                  2019-12-13T06:00:00.000Z,
                  2019-12-14T06:00:00.000Z,
                  2019-12-15T06:00:00.000Z,
                ]
              }
              key="TH"
            />
          </tbody>
        </Table>
      </styled.div>
    </Col>
  </Row>
</Container>
`;
